Add institutions app

Sam-Macpherson opened the pull request:
This change adds the institutions app, which will be our way of
tracking which universities/colleges are allowed to use our app.

Users who create accounts will need to choose their institution
from a static list, determined by our database of institutions.
The institutions each have a number of domains, which a user will
need to authenticate with when they sign up.

For example, the University of Waterloo has a few active domains:
uwaterloo.ca, edu.uwaterloo.ca.
If a student wants to sign up using their University of Waterloo
email, they will need to supply a uwaterloo.ca/edu.uwaterloo.ca
email.

This change also:
  * Adds the ID of listings to the model admin list view
  * Changes lists to tuples in the model admins (apparently tuples
    are more performant).
  * Adds the migration required for the listings app images to use local
    or aws bucket storage.

Sam-Macpherson commented:
Like a silly boy I forgot to add __pycache__ to the gitignore initally, so this change also removes those from the repo.

LzPwns reviewed (changes requested):
I think we have to blow up the whole repository.

Sam-Macpherson commented on listings/models/listing.py:
```diff
@@ -0,0 +1,48 @@
+        output = ''
+        if self.address:
+            output += f'{self.address} '
+        output += f'({self.id}) {str(self.start_date)} to ' \
```
Alright for now I can remove the ID field and if we ever need it down the line we can re-add it or something.

Sam-Macpherson commented on listings/models/listing.py:
```diff
@@ -0,0 +1,48 @@
+    end_date = DateField(
+        help_text=_('The date by which the subletter must leave.')
+    )
+    address = CharField(
```
Definitely something we'll make in the future.

Sam-Macpherson replied:
I think i'll add it as a property though, so we'll use it like `listing.term_length`

Sam-Macpherson commented on listings/models/listing.py:
```diff
@@ -0,0 +1,48 @@
+                  f'{str(self.end_date)} ' \
+                  f'({AccommodationType(self.accommodation_type).label})'
+        return output
+
```
Sure I'll add that ordering, I think start_date is proably the most apt.

Sam-Macpherson commented on subletshark/settings.py:
```diff
@@ -31,6 +31,7 @@
 # Application definition
 
 INSTALLED_APPS = [
+    'listings.apps.ListingsConfig',
```
Didn't know that, will change. The base app does not need to be added.

LzPwns commented on listings/models/room.py:
```diff
@@ -0,0 +1,47 @@
+        output = f'Bed: {BedType(self.bed_type).label}, ' \
+                 f'Ensuite: [{"✔" if self.ensuite else "✗"}] ' \
+                 f'Minifridge: [{"✔" if self.minifridge else "✗"}] ' \
+                 f'for listing: {self.listing_id} '
```
still a massive `__str__` but I'll let it slide because of those checkmarks

Sam-Macpherson replied:
I haven't changed it since the last PR.

LzPwns commented on listings/tests.py:
```diff
@@ -1,3 +1,66 @@
+            (datetime.date(2020, 1, 15), datetime.date(2020, 2, 28)),
+            # January 30 - March 1
+            (datetime.date(2020, 1, 30), datetime.date(2020, 3, 1)),
+            # May 15 - June 6, 1 month
```
nice

LzPwns commented on listings/tests.py:
```diff
@@ -1,3 +1,66 @@
+        for start_date, end_date in date_pairs:
+            self.listing.start_date = start_date
+            self.listing.end_date = end_date
+            self.listing.save()
```
I think you can just do `self.listing.update(start_date=start_date, end_date=end_date)`

Sam-Macpherson replied:
I was pretty sure that .update doesn't call the clean method and save does, but now that I've looked into it, it turns out that neither does .save(), so the clean I wrote was useless. I'll override the save() method to actually call our custom clean() method, and leave this as save() here, but I'll add the update_fields parameter.

LzPwns reviewed (changes requested):
I think you're gonna wanna update the tests once you change the term length calculation. Tests for 0 months, tests where the start date needs to be rounded (right now it's mostly using January 1st)

Sam-Macpherson commented on listings/models/listing.py:
```diff
@@ -0,0 +1,74 @@
+            months += 1
+        return months
+
+    def clean(self):
+        """Validate the fields before saving."""
+        super().clean()
+        if self.end_date > self.start_date:
+            raise ValidationError(_('The start date must be before the end '
+                                    'date of a sublet term.'))
+        if self.term_length < 1:
+            raise ValidationError(_('A sublet must be available for at least '
+                                    'one month.'))
```
I think you do not have to because no matter what the validation errors will get raised by super().clean() or not and then in line 63 and 66 as well, or not. Either way the necessary ValidationErrors are being raised here, but I will change it to make it more consistent with the django docs.

Sam-Macpherson commented:
Ignore the failing CI tests for now, I think they're failing because I need to set up postgres to run within them. The tests pass locally.

LzPwns commented on listings/models/listing.py:
```diff
@@ -0,0 +1,83 @@
+        """The length of the sublet term, in months."""
+        day_difference = (self.end_date - self.start_date).days
+        # 30.4375 is the average number of days in a month.
+        number_of_months = day_difference / 30.4375
```
should have this number in a constants file

Sam-Macpherson replied:
True.

LzPwns commented on listings/models/listing.py:
```diff
@@ -0,0 +1,83 @@
+        Example: January 14 -> January 1
+                 January 16 -> February 1
+        """
+        return (date + timedelta(days=16)).replace(day=1)
```
do we still need this method?

Sam-Macpherson replied:
I want to keep this method in because without it we'd be doing a calculation twice in 1 place, back to back, and this way we can make use of a docstring instead of comments to explain it.

Sam-Macpherson replied:
Also, in the next patch I'm adding that thing we talked about, "instead of using 16 we use half the # of days in the month in question", so that calculation will also take place here.

LzPwns commented on listings/models/listing.py:
```diff
@@ -0,0 +1,83 @@
+        day_difference = (self.end_date - self.start_date).days
+        # 30.4375 is the average number of days in a month.
+        number_of_months = day_difference / 30.4375
+        return int(number_of_months + 0.5)
```
even though this is a neat strat, you could just `round()`

Sam-Macpherson replied:
Lol yeah I guess I got too excited.

LzPwns reviewed (commented):
you're almost there champ

Sam-Macpherson committed:
Add institutions app

This change adds the institutions app, which will be our way of
tracking which universities/colleges are allowed to use our app.

Users who create accounts will need to choose their institution
from a static list, determined by our database of institutions.
The institutions each have a number of domains, which a user will
need to authenticate with when they sign up.

For example, the University of Waterloo has a few active domains:
uwaterloo.ca, edu.uwaterloo.ca.
If a student wants to sign up using their University of Waterloo
email, they will need to supply a uwaterloo.ca/edu.uwaterloo.ca
email.

This change also:
  * Adds the ID of listings to the model admin list view
  * Changes lists to tuples in the model admins (apparently tuples
    are more performant).

Sam-Macpherson committed:
Add basic serializer and view for institutions.

Sam-Macpherson committed:
Rename containers in docker-compose.

lazar-vukoje commented on institutions/models.py:
```diff
@@ -0,0 +1,43 @@
+    )
+
+    def __str__(self):
+        """Human readable representation of the institution."""
```
I think we might be going a little overboard with all these comments. I don't mind comments on the model, but we don't really need  """Human readable representation of the ___.""" under every str().

Sam-Macpherson replied:
I'm definitely going to keep writing doc strings in accordance with good python practice.

lazar-vukoje commented on institutions/serializers.py:
```diff
@@ -0,0 +1,18 @@
+
+    class Meta:
+        model = Institution
+        exclude = ('id', )
```
why exclude id?

Sam-Macpherson replied:
I gave this one some thought, and I decided that since an institution is highly unique and unique by name, an ID is a useless field to know. We can always identify an institution by its name, and would be better served by searching for it by name on the admin page anyway, so I applied the same thinking to in the serialized output. An ID field for this model would just be clutter.

Sam-Macpherson replied:
Even in a wacky case where maybe there are two University of Waterloo in the world, we would need a way to distinguish them anyway, maybe by address or a new location field down the line, an id isn't worth it I think for the returned objects.

lazar-vukoje commented on institutions/views.py:
```diff
@@ -0,0 +1,17 @@
+
+    def list(self, request, **kwargs):
+        serializer = InstitutionSerializer(self.queryset, many=True)
+        return Response(serializer.data)
```
ModelViewSet provides all the standard actions, including list(). You don't need to override it since you're not doing any custom logic.

Sam-Macpherson replied:
Indeed. I just wrote this here to remind me how it worked.

lazar-vukoje commented on listings/views.py:
```diff
@@ -7,8 +10,9 @@
     serializer_class = ListingSerializer
+    permission_classes = [IsAuthenticated]
 
     def list(self, request, **kwargs):
```
same for this guy

Sam-Macpherson replied:
Ye

Sam-Macpherson replied:
I guess I can take them out for now. They'll get removed eventually anyway once we write more complicated views.

Sam-Macpherson committed:
Remove extra list methods on view sets.
